presto-main/src/main/java/com/facebook/presto/operator/aggregation/AreaUnderROCCurveAggregation.java

```diff
@@ -0,0 +1,217 @@
+ * limitations under the License.
+ */
+package com.facebook.presto.operator.aggregation;
+
+import com.facebook.presto.bytecode.DynamicClassLoader;
+import com.facebook.presto.metadata.BoundVariables;
+import com.facebook.presto.metadata.FunctionRegistry;
+import com.facebook.presto.metadata.SqlAggregationFunction;
+import com.facebook.presto.operator.aggregation.state.MultiKeyValuePairStateSerializer;
+import com.facebook.presto.operator.aggregation.state.MultiKeyValuePairsState;
+import com.facebook.presto.operator.aggregation.state.MultiKeyValuePairsStateFactory;
+import com.facebook.presto.spi.block.Block;
+import com.facebook.presto.spi.block.BlockBuilder;
+import com.facebook.presto.spi.type.Type;
+import com.facebook.presto.spi.type.TypeManager;
+import com.google.common.collect.ImmutableList;
+
+import java.lang.invoke.MethodHandle;
+import java.util.List;
+import java.util.stream.IntStream;
+
+import static com.facebook.presto.operator.aggregation.AggregationMetadata.ParameterMetadata.ParameterType.BLOCK_INDEX;
+import static com.facebook.presto.operator.aggregation.AggregationMetadata.ParameterMetadata.ParameterType.BLOCK_INPUT_CHANNEL;
+import static com.facebook.presto.operator.aggregation.AggregationMetadata.ParameterMetadata.ParameterType.STATE;
+import static com.facebook.presto.operator.aggregation.AggregationUtils.generateAggregationName;
+import static com.facebook.presto.spi.type.BooleanType.BOOLEAN;
+import static com.facebook.presto.spi.type.DoubleType.DOUBLE;
+import static com.facebook.presto.util.Reflection.methodHandle;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+
+public class AreaUnderROCCurveAggregation
+        extends SqlAggregationFunction
+{
+    public static final AreaUnderROCCurveAggregation AUC = new AreaUnderROCCurveAggregation();
+    public static final String NAME = "area_under_roc_curve";
+    private static final MethodHandle
+            INPUT_FUNCTION = methodHandle(AreaUnderROCCurveAggregation.class, "input", MultiKeyValuePairsState.class, Block.class, Block.class, int.class);
+    private static final MethodHandle COMBINE_FUNCTION = methodHandle(AreaUnderROCCurveAggregation.class, "combine", MultiKeyValuePairsState.class, MultiKeyValuePairsState.class);
+    private static final MethodHandle OUTPUT_FUNCTION = methodHandle(AreaUnderROCCurveAggregation.class, "output", MultiKeyValuePairsState.class, BlockBuilder.class);
+
+    protected AreaUnderROCCurveAggregation()
+    {
+        super(NAME,
+                ImmutableList.of(),
+                ImmutableList.of(),
+                DOUBLE.getTypeSignature(),
+                ImmutableList.of(BOOLEAN.getTypeSignature(), DOUBLE.getTypeSignature()));
+    }
+
+    @Override
+    public String getDescription()
+    {
+        return "Returns AUC";
+    }
+
+    @Override
+    public InternalAggregationFunction specialize(BoundVariables boundVariables, int arity, TypeManager typeManager, FunctionRegistry functionRegistry)
```

Comment:
This doesn't need to be a parametric function. The types of its arguments are well-known. Take a look at how functions like sum, avg, corr, etc. are implemented.

For the state, you'll want to have a custom object that keeps track of the labels and scores similar to how the Histogram aggregation does. There's a pending PR (still work in progress) to optimize how Histogram stores its state when dealing with grouped aggregations to reduce per-group overhead. You may want to look into that, too: https://github.com/prestodb/presto/pull/9037